Nullable reference types in F# (#44047)
dotnet/docs: docs/fsharp/style-guide/component-design-guidelines.md

    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj) =
    if isNull arg then nullArg argName
    else ()
```

Starting with F# 9, you can leverage the new `| null` [syntax](../language-reference/values/null-values.md#null-values-starting-with-f-9) to make the compiler indicate possible null values and where they need handling:

```fsharp
let checkNonNull argName (arg: obj | null) =
    match arg with
    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj | null) =
    if isNull arg then nullArg argName 
    else ()
```

In F# 9, the compiler emits a warning when it detects that a possible null value is not handled:

```fsharp
let printLineLength (s: string) =
    printfn "%i" s.Length

let readLineFromStream (sr: System.IO.StreamReader) =
    // `ReadLine` may return null here - when the stream is finished
    let line = sr.ReadLine()
    // nullness warning: The types 'string' and 'string | null'
    // do not have equivalent nullability
    printLineLength line
```

> These warnings should be addressed using idiomatic F# [null pattern](../language-reference/pattern-matching.md#null-pattern) via pattern matching:

Reviewer: ```suggestion
These warnings should be addressed using F# [null pattern](../language-reference/pattern-matching.md#null-pattern) in matching:
```

("pattern" is already included in the hyperlink text)

    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj) =
    if isNull arg then nullArg argName
    else ()
```

Starting with F# 9, you can leverage the new `| null` [syntax](../language-reference/values/null-values.md#null-values-starting-with-f-9) to make the compiler indicate possible null values and where they need handling. The above code will then start producing warnings:

```fsharp
let checkNonNull argName (arg: obj) =
    match arg with
    // nullness warning: the type 'obj' does not support 'null'
    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj) =
    // nullness warning: the type 'obj' does not support 'null'
    if isNull arg then nullArg argName 
    else ()
```

> To address the warnings, you need to explicitly specify `null` as a possible argument value:

Reviewer: This set of examples is not very idiomatic, neither the "bad" ones nor the good ones.
The built-in `nullArgCheck` function handles this with two important properties added:
- It does work with type inference and type annotation is not needed
- It uses shadowing to return a non-null value for the binding, so that subsequent code work with a value without nulls.

Author: Correct, rewrote the code using `nullArgCheck`. Also added some notes about `nullArgCheck` and `nonNull` to the conventions page.

    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj) =
    if isNull arg then nullArg argName
    else ()
```

Starting with F# 9, you can leverage the new `| null` [syntax](../language-reference/values/null-values.md#null-values-starting-with-f-9) to make the compiler indicate possible null values and where they need handling. The above code will then start producing warnings:

```fsharp

> let checkNonNull argName (arg: obj) =

Reviewer: The code does not relate to the text.
Also the text mentiones "above code" but should say "below code" ?

More importantly, the text and indeed the main motivation is to warn on possible nullness exceptions.
This code sample rather shows a "useless null check", since the code is overly defensive in handling null where it cannot be present.

Better starting example is (string|null).Length and a warning for it.

Author: The point here was to say that the same (above) code would start producing warnings in F# 9.
But if that's not too clear, that should be indeed rewritten.

I removed some text and examples to decrease the content load here.

    match arg with
    // nullness warning: the type 'obj' does not support 'null'
    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj) =
    // nullness warning: the type 'obj' does not support 'null'

>     if isNull arg then nullArg argName 

Reviewer: Does not relate to the text.

Author: Addressed.

    else ()
```

> To address the warnings, you could explicitly specify `null` as a possible argument value:

Reviewer: This needs rework based on how the first example is redone.
Right now, it shows the wrong incentive.

If e.g. the first example shows a warning for (string|null).Length, that this example should show how to handle it.

Author: I removed and reworked the text, take a look if it's more consistent and helpful now.

```fsharp
let checkNonNull argName (arg: obj | null) =
    match arg with
    | null -> nullArg argName
    | _ -> ()

let checkNonNull' argName (arg: obj | null) =
    if isNull arg then nullArg argName 
    else ()
```

> Yet, the most succinct way to achieve this is to use the new builtin `nullArgCheck` function:

Reviewer: This does not fit here, `nullArgCheck` is a throwing validation, more of a last resort.
The motivating examples should start be doing the right thing by handling nulls without failing, this should not be demonstrated as a succint way.

Author: Fair, `nullArgCheck` is mentioned elsewhere and doesn't need to be here.

```fsharp
let checkNonNull'' argName arg =    // `arg` is inferred to be `obj | null`
    nullArgCheck argName arg
    |> ignore
```

> The `nullArgCheck` function is also handy when you need to sanitize the input and keep it in case it's not null:

Reviewer: This can stay here, but should be done at the bottom and very explicitely written as a validation routine **that throws**, i.e. more of a last resort.

Author: `nullArgCheck` is mentioned in "conventions" which explicitly mentioned that, so indeed doesn't need to be here.

```fsharp
let processNullableList l =       
    // `l` is inferred as `int list | null`
    let l = nullArgCheck (nameof l) l
    // now, `l` is inferred as `int list` and is safe to use
    l |> List.map (fun x -> x + 1)
```

On the other hand, you'll get a warning if the compiler detects that a possible null value is not handled:

```fsharp
let printLineLength (s: string) =
    printfn "%i" s.Length

> let readLineFromStream (sr: System.IO.StreamReader) =

Reviewer: This is a good example and might as well come first.
It should say where the nulls comes from - it is part of the annotation for `.ReadLine()`

Author: Right, added a note.

    let line = sr.ReadLine()
    // nullness warning: The types 'string' and 'string | null'
    // do not have equivalent nullability
    printLineLength line
```

These warnings should be addressed using idiomatic F# [null pattern](../language-reference/pattern-matching.md#null-pattern) via pattern matching:

```fsharp
let printLineLength (s: string) =
    printfn "%i" s.Length

let readLineFromStream (sr: System.IO.StreamReader) =
    let line = sr.ReadLine()
    match line with

>     | Null -> ()

Author: Addressed, thanks.
